homework_solution.md

```diff
@@ -0,0 +1,474 @@
+# Part 1: Missing Station Data
+
+1. _**Write a query that fills out this table. Try your best to pick the correct station name for each ID. You may have to make some manual choices or editing based on the inconsistencies we've found. Do try to pick the correct name for each station ID based on how popular it is in the trip data.**_
+
+   ### Steps
+
+   1. Create an index on `from_station_id` and `to_station_id` on `trips` table to help query on those
+
+      ```
+      CREATE INDEX idx_from_station ON trips USING btree (from_station_id);
+      ```
+
+      ```
+      CREATE INDEX idx_to_station ON trips USING btree (to_station_id);
+      ```
+
+   2. Delete the exist `stations` table and recreate it with serial ids, so we can rely on auto-increment. (We could try and alter the table, but that seems to be an involved process, and our table being currently empty, it's easier to just recreate it)
+
+      ```
+      DROP TABLE stations;
+      ```
+
+      ```
+      CREATE TABLE stations
+      (
+         id SERIAL,
+         name VARCHAR(100)
+      );
+      ```
+
+      The effect of this can be perceived by running `SELECT nextval(pg_get_serial_sequence('stations', 'id'));`.
+      On the old table, that sequence, though present, wouldn't return a value. Now it does.
+
+   3. Get the first name for stations, assuming that as the correct one where there's more than one, combining the `from` and `to` column results
+
+      ```
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+      FROM trips
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id
+      UNION
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+      FROM trips
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id;
+      ```
+
+   4. Insert into the `stations` table the result of our query
+
+      ```
+      INSERT INTO stations (
+         SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+         FROM trips
+         WHERE from_station_id IS NOT NULL
+         GROUP BY from_station_id
+         UNION
+         SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+         FROM trips
+         WHERE from_station_id IS NOT NULL
+         GROUP BY from_station_id
+      );
+      ```
+
+   5. Adjust the `id` sequence to be the highest `id` from our `INSERT` above
+
+      ```
+      SELECT setval(pg_get_serial_sequence('stations', 'id'), (SELECT MAX(id) FROM stations));
+      ```
+
+   6. Finally, add entries for records where a name exists, but the id is null, relying on the seqquence to generate an id. This has to be done in 2 steps, so we can exclude the records from the first insertions when we run the second one, avoiding duplicates.
+
+      ```
+      INSERT INTO stations (
+         SELECT nextval(pg_get_serial_sequence('stations', 'id')), from_station_name
```

Comment:
[question]
Why do you need to select this explicitly? Shouldn't this be handled by autoincrement?

Reply:
The way I understand this, it would, if my query were something like `INSERT INTO stations (name) values ('aStationName').

Running a query for multiple inserts, I couldn't get it to work (maybe there's a way, but I couldn't find it), so I had to the explicit call to the `nextval`.

Reply:
Huh. That's weird. I wonder why that is.  SQL isn't my strongest skill so I'm not sure how much help I can provide.  I'd encourage you to ask the other mentors about it though - there's gotta be a way to do autoincrementing primary keys for a bulk insert like this

Reply:
Will do!